### ISSUE TITLE
add, change, then delete

### ISSUE THREAD

**Issue by buckle2000** [bakpakin/tiny-ecs#8]
refer to tiny.lua Line 643-747
I think the correct order should be: add, change, then delete because if I do this:
- Add a entity
- Remove it immediately without call to `world:refresh`

This will NOT trigger a system's onRemove callback, which does not make sense.
Also, I think it is better not to use `filter` in the "add entities" procedure, but to add entities into `entitiesToChange` even if they are newly added.
This requires changes to:
- `tiny_manageEntities`, of course
- `tiny.addEntity`

**Comment by bakpakin** [comment]
After going through the code, it seems that the need for separate loops to add and change entities is no longer there. I was able to merge the adding and changing into one loop, which simplified the code a bit as well. Thanks for pointing this out! Because changing/adding  now all happens before remove, your desired behavior should work as expected.

As for the bit about filters, I believe the new code addresses that issue with a slightly different solution.

I've just pushed a change to master - check it out and see if it solves your problems. I want to write some test-cases as well if it helps

**Comment by buckle2000** [comment]
Well, I don't think checking whether an entity is already added every time it changes is a good idea.
Although **changing components** is not a frequent kind of event and will not make program much slower.
Instead, when an entity is added, add it to both `e2a` and `e2c`, and change `tiny_manageEntities` accordingly.

One more thing, 
is `world.entityCount` necessary?
is it equal to `#world.entityList`?

**Comment by bakpakin** [comment]
As you say, changing components does not happen often, and the performance difference would be negligible. Also, because you use `tiny.addEntity` to indicate that components have changed as well as to add new entities, **checking if the entity has already been added has to happen anyways**. Whether it happens in `tiny.addEntities` or in `tiny_manageEntities` is the only change in the new code. These two separate uses for `tiny.addEntity` are combinded into one function to keep the API simple. 

But yeah, `world.entityCount` should always equal `#world.entityList`. The separate entity count is vestigial; it comes from an early version where entities were not kept in an ordered table, but were in a set.

While going through this, another old design decision that is no longer necessary is keeping separate entity-to-index and index-to-entity tables. Assuming that entities are not numbers, these tables could be merged. I've made this change, as well as removing `world.entityCount`, and pushed up a new version.
